src/ca.c

```diff
@@ -2736,625 +2419,770 @@ static int ciplus_app_upgr_message(void *arg, uint8_t slot_id,
-    int fd;
-    fd = open("/etc/ssl/certs/root.pem", O_RDONLY);
-    if (fd < 0)
-        d->force_ci = 1;
-    else
-        close(fd);
-    if (d->force_ci) {
-        resource = resource_ids_ci;
-        resource_count = resource_ids_ci_count;
-    }
-
-    LOG("%02x:%s  resource_count %i %s", slot_id, __func__, resource_count,
-        d->force_ci ? "CI MODE" : "CI+ MODE");
-
-    if (en50221_app_rm_reply(d->rm_resource, session_number, resource_count,
-                             resource)) {
-        LOG("%02x:Failed to send reply to ENQ", slot_id);
+int APP_DateTime_handler(struct ca_device *d, int resource, uint8_t *buffer,
+                         int len) {
+    switch (resource) {
+    case TAG_DATE_TIME_ENQUIRY:
+        if (buffer)
+            d->datetime_response_interval = *buffer;
+        break;
+    default:
+        LOG("unexpected tag in %s (0x%x)", __FUNCTION__, resource);
     }
-
+    ca_send_datetime(d);
     return 0;
 }
 
-static int ca_rm_reply_callback(void *arg, uint8_t slot_id,
-                                uint16_t session_number,
-                                uint32_t resource_id_count,
-                                uint32_t *_resource_ids) {
-    ca_device_t *d = arg;
-    LOG("%02x:%s resource_count %i", slot_id, __func__, resource_id_count);
+int ca_app_mmi_answ(ca_device_t *d, uint8_t answ_id, uint8_t *text,
+                    uint32_t text_count) {
+    uint8_t buf[10 + text_count];
+    int len = 1 + text_count;
 
-    uint32_t i;
-    for (i = 0; i < resource_id_count; i++) {
-        LOG("  CAM provided resource id: %08x", _resource_ids[i]);
-    }
+    buf[0] = answ_id;
 
-    if (en50221_app_rm_changed(d->rm_resource, session_number)) {
-        LOG("%02x:Failed to send REPLY", slot_id);
+    if (text_count > 0) {
+        memcpy(buf + 1, text, text_count);
     }
+    return ca_write_apdu(d, TAG_ANSWER, buf, len);
+}
 
-    return 0;
+int ca_app_mmi_menu_answ(ca_device_t *d, uint8_t answ_id) {
+    return ca_write_apdu(d, TAG_MENU_ANSWER, &answ_id, 1);
 }
 
-static int ca_rm_changed_callback(void *arg, uint8_t slot_id,
-                                  uint16_t session_number) {
-    ca_device_t *d = arg;
-    LOG("%02x:%s", slot_id, __func__);
+int ca_app_mmi_close(ca_device_t *d, uint8_t answ_id) {
+    return ca_write_apdu(d, TAG_CLOSE_MMI, &answ_id, 1);
+}
+
+int APP_MMI_handler(struct ca_device *d, int resource, uint8_t *buffer,
+                    int len) {
+    hexdump("Got MMI: ", buffer, len);
+    switch (resource) {
+    case TAG_DISPLAY_CONTROL: {
+        int cmd_id = buffer[0];
+        int mmi_mode = buffer[1];
+        if (cmd_id == MMI_DISPLAY_CONTROL_CMD_ID_SET_MMI_MODE) {
 
-    if (en50221_app_rm_enq(d->rm_resource, session_number)) {
-        LOG("%02x:Failed to send ENQ", slot_id);
+            LOG("mmi display ctl cb received for CA %u session_num %u "
+                "cmd_id 0x%02x mmi_mode %u",
+                d->id, d->session_number, cmd_id, mmi_mode);
```

Comment:
Here for example it seems like `d->session_number` isn't actually the session number for the MMI resource. Maybe this field should be removed from `ca_device_t` completely to avoid confusion?

Reply:
we have 2 choices:
- use d->sesion_number (or rename it)
- carry session_number as argument across all the method as part of the CAM request
WDYT ?